File: 06_Shared_Components/controllers/application-controller.ts
/**
 * KPRFlow Enterprise - Application Controller
 * Adapted from bramsray2 backend for cross-platform use
 */

import { 
  Application, 
  ApplicationStatus, 
  ApplicationCreateRequest, 
  ApplicationUpdateRequest,
  ApplicationFilters,
  ApplicationResponse,
  ApplicationListResponse,
  ApplicationStatistics,
  SupabaseQueryBuilder
} from '../types/supabase-types'
import { ApiResponse, PaginatedResponse } from '../types/common-types'
import { supabaseClient } from '../api_clients/supabase-client'
import { notificationService } from '../services/notification-service'
import { DateCalculator } from '../utils/date-calculator'

export class ApplicationController {
  private static instance: ApplicationController

  private constructor() {}

  static getInstance(): ApplicationController {
    if (!ApplicationController.instance) {
      ApplicationController.instance = new ApplicationController()
    }
    return ApplicationController.instance
  }

  /**
   * Get all applications with optional filters
   */
  async getAllApplications(
    filters?: ApplicationFilters,
    page: number = 1,
    limit: number = 20
  ): Promise<ApplicationListResponse> {
    try {
      const client = supabaseClient.getClient()
      let query = client.from('kpr_applications') as any

      // Apply filters
      if (filters) {
        if (filters.status && filters.status.length > 0) {
          query = query.in('status', filters.status)
        }
        if (filters.customer_id) {
          query = query.eq('customer_id', filters.customer_id)
        }
        if (filters.unit_id) {
          query = query.eq('unit_id', filters.unit_id)
        }
        if (filters.bank_id) {
          query = query.eq('bank_id', filters.bank_id)
        }
        if (filters.date_from) {
          query = query.gte('created_at', filters.date_from)
        }
        if (filters.date_to) {
          query = query.lte('created_at', filters.date_to)
        }
        if (filters.search) {
          query = query.or(`customer.name.ilike.%${filters.search}%,unit.block.ilike.%${filters.search}%`)
        }
      }

      // Apply pagination
      const offset = (page - 1) * limit
      query = query.range(offset, offset + limit - 1)

      // Get count
      const { count } = await client
        .from('kpr_applications')
        .select('*', { count: 'exact', head: true })

      // Execute query
      const { data, error } = await query

      if (error) {
        return {
          data: undefined,
          error: error.message,
          count: 0,
          has_more: false,
          page,
          limit
        }
      }

      const applicationsWithAging = (data || []).map(app => ({
        ...app,
        aging_days: DateCalculator.calculateDaysSince(app.created_at),
        sla_info: this.calculateSLAInfo(app)
      }))

      return {
        success: true,
        data: applicationsWithAging,
        count: applicationsWithAging.length,
        page,
        limit,
        total: count || 0,
        has_more: (count || 0) > offset + limit
      }
    } catch (error: any) {
      console.error('Error fetching applications:', error)
      return {
        success: false,
        data: [],
        count: 0,
        error: error.message
      }
    }
  }

  /**
   * Get application by ID
   */
  async getApplicationById(id: string): Promise<ApplicationResponse<Application>> {
    try {
      const client = supabaseClient.getClient()
      
      const { data, error } = await client
        .from('kpr_applications')
        .select(`
          *,
          customers:customer_id(*),
          units:unit_id(*),
          banks:bank_id(*),
          application_documents(id, document_type, file_name, file_url, uploaded_at, verified),
          application_status_history(id, from_status, to_status, changed_by, changed_at, notes)
        `)
        .eq('id', id)
        .single()

      if (error) throw error

      // Calculate SLA info
      const slaInfo = DateCalculator.getSLAStatus(
        data.created_at,
        new Date(),
        this.getSLADaysByStatus(data.status)
      )

      const application = {
        ...data,
        sla_info: slaInfo,
        aging_days: DateCalculator.calculateAging(data.created_at).days
      }

      return {
        success: true,
        data: application
      }
    } catch (error: any) {
      console.error('Error fetching application:', error)
      return {
        success: false,
        error: error.message
      }
    }
  }

  /**
   * Create new application
   */
  async createApplication(request: ApplicationCreateRequest): Promise<ApplicationResponse<Application>> {
    try {
      const client = supabaseClient.getClient()

      // Validate customer and unit availability
      const validation = await this.validateApplicationCreation(request)
      if (!validation.isValid) {
        return {
          success: false,
          error: validation.errors.join(', ')
        }
      }

      // Create application
      const applicationData = {
        ...request,
        status: 'PENDING' as ApplicationStatus,
        created_at: new Date().toISOString(),
        updated_at: new Date().toISOString()
      }

      const { data, error } = await client
        .from('kpr_applications')
        .insert([applicationData])
        .select(`
          *,
          customers:customer_id(name, email, phone),
          units:unit_id(unit_type, block, number, price)
        `)
        .single()

      if (error) throw error

      // Send notification
      await this.sendStatusChangeNotification(data, null, 'PENDING')

      // Update unit status if needed
      await this.updateUnitStatus(request.unit_id, 'RESERVED')

      return {
        success: true,
        data,
        message: 'Aplikasi KPR berhasil dibuat'
      }
    } catch (error: any) {
      console.error('Error creating application:', error)
      return {
        success: false,
        error: error.message
      }
    }
  }

  /**
   * Update application
   */
  async updateApplication(id: string, request: ApplicationUpdateRequest): Promise<ApplicationResponse<Application>> {
    try {
      const client = supabaseClient.getClient()

      // Get current application
      const currentApp = await this.getApplicationById(id)
      if (!currentApp.success || !currentApp.data) {
        return {
          success: false,
          error: 'Aplikasi tidak ditemukan'
        }
      }

      // Validate status change
      if (request.status && request.status !== currentApp.data.status) {
        const validation = await this.validateStatusChange(
          currentApp.data.status,
          request.status,
          currentApp.data
        )
        if (!validation.isValid) {
          return {
            success: false,
            error: validation.errors.join(', ')
          }
        }
      }

      // Update application
      const updateData = {
        ...request,
        updated_at: new Date().toISOString()
      }

      const { data, error } = await client
        .from('kpr_applications')
        .update(updateData)
        .eq('id', id)
        .select(`
          *,
          customers:customer_id(name, email, phone),
          units:unit_id(unit_type, block, number, price),
          banks:bank_id(name, contact_person)
        `)
        .single()

      if (error) throw error

      // Log status change
      if (request.status && request.status !== currentApp.data.status) {
        await this.logStatusChange(
          id,
          currentApp.data.status,
          request.status,
          request.notes
        )

        // Send notification
        await this.sendStatusChangeNotification(
          data,
          currentApp.data.status,
          request.status
        )

        // Update unit status if needed
        await this.handleStatusChangeUnitUpdate(request.unit_id, request.status)
      }

      return {
        success: true,
        data,
        message: 'Aplikasi berhasil diperbarui'
      }
    } catch (error: any) {
      console.error('Error updating application:', error)
      return {
        success: false,
        error: error.message
      }
    }
  }

  /**
   * Delete application
   */
  async deleteApplication(id: string): Promise<ApplicationResponse> {
    try {
      const client = supabaseClient.getClient()

      // Get application before deletion
      const currentApp = await this.getApplicationById(id)
      if (!currentApp.success || !currentApp.data) {
        return {
          success: false,
          error: 'Aplikasi tidak ditemukan'
        }
      }

      // Check if can be deleted
      if (!this.canDeleteApplication(currentApp.data.status)) {
        return {
          success: false,
          error: 'Aplikasi tidak dapat dihapus pada status ini'
        }
      }

      // Delete application
      const { error } = await client
        .from('kpr_applications')
        .delete()
        .eq('id', id)

      if (error) throw error

      // Update unit status
      await this.updateUnitStatus(currentApp.data.unit_id, 'AVAILABLE')

      return {
        success: true,
        message: 'Aplikasi berhasil dihapus'
      }
    } catch (error: any) {
      console.error('Error deleting application:', error)
      return {
        success: false,
        error: error.message
      }
    }
  }

  /**
   * Get applications by status
   */
  async getApplicationsByStatus(
    status: ApplicationStatus,
    page: number = 1,
    limit: number = 20
  ): Promise<ApplicationListResponse> {
    return this.getAllApplications({ status }, page, limit)
  }

  /**
   * Get application statistics
   */
  async getApplicationStatistics(): Promise<ApplicationResponse<ApplicationStatistics>> {
    try {
      const client = supabaseClient.getClient()

      // Get total applications by status
      const { data: statusData, error: statusError } = await client
        .from('kpr_applications')
        .select('status')
        .not('status', 'eq', 'CANCELLED')

      if (statusError) throw statusError

      const statusCounts = statusData?.reduce((acc, app) => {
        acc[app.status] = (acc[app.status] || 0) + 1
        return acc
      }, {} as Record<ApplicationStatus, number>) || {}

      // Get applications by bank
      const { data: bankData, error: bankError } = await client
        .from('kpr_applications')
        .select('banks!inner(name)')
        .not('bank_id', 'is', null)

      if (bankError) throw bankError

      const bankCounts = bankData?.reduce((acc, app) => {
        const bankName = app.banks.name
        acc[bankName] = (acc[bankName] || 0) + 1
        return acc
      }, {} as Record<string, number>) || {}

      // Get monthly statistics
      const currentMonth = new Date().toISOString().substring(0, 7)
      const { data: monthlyData, error: monthlyError } = await client
        .from('kpr_applications')
        .select('created_at')
        .like('created_at', `${currentMonth}%`)

      if (monthlyError) throw monthlyError

      const monthlyStats = monthlyData?.reduce((acc, app) => {
        const day = app.created_at.substring(8, 10)
        acc[day] = (acc[day] || 0) + 1
        return acc
      }, {} as Record<string, number>) || {}

      // Calculate approval and rejection rates
      const total = Object.values(statusCounts).reduce((sum, count) => sum + count, 0)
      const approved = statusCounts['APPROVED'] || 0
      const rejected = statusCounts['REJECTED'] || 0

      const statistics: ApplicationStatistics = {
        total,
        by_status: statusCounts,
        by_bank: bankCounts,
        by_month: monthlyStats,
        average_processing_time: 0, // Would need more complex calculation
        approval_rate: total > 0 ? (approved / total) * 100 : 0,
        rejection_rate: total > 0 ? (rejected / total) * 100 : 0
      }

      return {
        success: true,
        data: statistics
      }
    } catch (error: any) {
      console.error('Error fetching statistics:', error)
      return {
        success: false,
        error: error.message
      }
    }
  }

  /**
   * Bulk update applications
   */
  async bulkUpdateApplications(
    ids: string[],
    updateData: ApplicationUpdateRequest
  ): Promise<ApplicationResponse<Application[]>> {
    try {
      const client = supabaseClient.getClient()

      const { data, error } = await client
        .from('kpr_applications')
        .update({
          ...updateData,
          updated_at: new Date().toISOString()
        })
        .in('id', ids)
        .select(`
          *,
          customers:customer_id(name, email, phone),
          units:unit_id(unit_type, block, number, price)
        `)

      if (error) throw error

      return {
        success: true,
        data: data || [],
        message: `${ids.length} aplikasi berhasil diperbarui`
      }
    } catch (error: any) {
      console.error('Error bulk updating applications:', error)
      return {
        success: false,
        error: error.message
      }
    }
  }

  // Private helper methods
  private async validateApplicationCreation(request: ApplicationCreateRequest): Promise<{ isValid: boolean; errors: string[] }> {
    const errors: string[] = []

    // Validate customer exists
    const customer = await supabaseClient.getClient()
      .from('customers')
      .select('id')
      .eq('id', request.customer_id)
      .single()

    if (customer.error) {
      errors.push('Customer tidak ditemukan')
    }

    // Validate unit exists and is available
    const unit = await supabaseClient.getClient()
      .from('units')
      .select('id, status')
      .eq('id', request.unit_id)
      .single()

    if (unit.error) {
      errors.push('Unit tidak ditemukan')
    } else if (unit.data.status !== 'AVAILABLE') {
      errors.push('Unit tidak tersedia')
    }

    // Validate bank if provided
    if (request.bank_id) {
      const bank = await supabaseClient.getClient()
        .from('banks')
        .select('id')
        .eq('id', request.bank_id)
        .single()

      if (bank.error) {
        errors.push('Bank tidak ditemukan')
      }
    }

    return {
      isValid: errors.length === 0,
      errors
    }
  }

  private async validateStatusChange(
    fromStatus: ApplicationStatus,
    toStatus: ApplicationStatus,
    application: Application
  ): Promise<{ isValid: boolean; errors: string[] }> {
    const errors: string[] = []

    // Define valid status transitions
    const validTransitions: Record<ApplicationStatus, ApplicationStatus[]> = {
      'PENDING': ['DOCUMENT_COLLECTION', 'CANCELLED'],
      'DOCUMENT_COLLECTION': ['BANK_SUBMISSION', 'CANCELLED'],
      'BANK_SUBMISSION': ['BANK_VERIFICATION', 'CANCELLED'],
      'BANK_VERIFICATION': ['BANK_APPROVAL', 'REJECTED', 'CANCELLED'],
      'BANK_APPROVAL': ['APPROVED', 'CANCELLED'],
      'APPROVED': ['COMPLETED', 'CANCELLED'],
      'COMPLETED': ['BAST_COMPLETED'],
      'BAST_COMPLETED': [],
      'REJECTED': [],
      'CANCELLED': []
    }

    const allowedStatuses = validTransitions[fromStatus] || []
    if (!allowedStatuses.includes(toStatus)) {
      errors.push(`Status change dari ${fromStatus} ke ${toStatus} tidak diizinkan`)
    }

    return {
      isValid: errors.length === 0,
      errors
    }
  }

  private async logStatusChange(
    applicationId: string,
    fromStatus: ApplicationStatus,
    toStatus: ApplicationStatus,
    notes?: string
  ): Promise<void> {
    try {
      const client = supabaseClient.getClient()
      
      await client.from('application_status_history').insert({
        application_id: applicationId,
        from_status: fromStatus,
        to_status: toStatus,
        changed_by: 'system', // Would get from auth context
        changed_at: new Date().toISOString(),
        notes
      })
    } catch (error) {
      console.error('Error logging status change:', error)
    }
  }

  private async sendStatusChangeNotification(
    application: Application,
    fromStatus: ApplicationStatus | null,
    toStatus: ApplicationStatus
  ): Promise<void> {
    try {
      const notification = notificationService.createFromTemplate(
        'status_change',
        {
          customer_name: application.customer?.name,
          application_id: application.id,
          unit_block: application.unit?.block,
          unit_number: application.unit?.number,
          new_status: toStatus,
          previous_status: fromStatus
        },
        application.customer_id,
        application.id
      )

      if (notification) {
        await notificationService.sendNotification(notification)
      }
    } catch (error) {
      console.error('Error sending status change notification:', error)
    }
  }

  private async updateUnitStatus(unitId: string, status: string): Promise<void> {
    try {
      const client = supabaseClient.getClient()
      await client.from('units')
        .update({ status, updated_at: new Date().toISOString() })
        .eq('id', unitId)
    } catch (error) {
      console.error('Error updating unit status:', error)
    }
  }

  private async handleStatusChangeUnitUpdate(unitId: string, status: ApplicationStatus): Promise<void> {
    const unitStatusMap: Record<ApplicationStatus, string> = {
      'PENDING': 'RESERVED',
      'DOCUMENT_COLLECTION': 'RESERVED',
      'BANK_SUBMISSION': 'RESERVED',
      'BANK_VERIFICATION': 'RESERVED',
      'BANK_APPROVAL': 'RESERVED',
      'APPROVED': 'SOLD',
      'COMPLETED': 'SOLD',
      'BAST_COMPLETED': 'SOLD',
      'REJECTED': 'AVAILABLE',
      'CANCELLED': 'AVAILABLE'
    }

    const unitStatus = unitStatusMap[status]
    if (unitStatus) {
      await this.updateUnitStatus(unitId, unitStatus)
    }
  }

  private canDeleteApplication(status: ApplicationStatus): boolean {
    return ['PENDING', 'DOCUMENT_COLLECTION'].includes(status)
  }

  private getSLADaysByStatus(status: ApplicationStatus): number {
    const slaDays: Record<ApplicationStatus, number> = {
      'PENDING': 3,
      'DOCUMENT_COLLECTION': 7,
      'BANK_SUBMISSION': 2,
      'BANK_VERIFICATION': 14,
      'BANK_APPROVAL': 3,
      'APPROVED': 30,
      'COMPLETED': 0,
      'BAST_COMPLETED': 0,
      'REJECTED': 0,
      'CANCELLED': 0
    }
    return slaDays[status] || 30
  }

  private getSLAStatus(createdDate: string, status: ApplicationStatus): 'on_track' | 'warning' | 'overdue' {
    const slaDays = this.getSLADaysByStatus(status)
    const slaInfo = DateCalculator.getSLAStatus(createdDate, new Date(), slaDays)
    return slaInfo.status === 'on_track' ? 'on_track' : slaInfo.status
  }
}

// Export singleton instance
export const applicationController = ApplicationController.getInstance()

File: 01_Mobile_APK_Android/app/src/main/java/com/kprflow/android/shared/ApplicationController.kt
package com.kprflow.android.shared

import android.util.Log
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.map

// Application Controller - Adapted from 06_Shared_Components
class ApplicationController private constructor() {
    
    companion object {
        private const val TAG = "ApplicationController"
        
        @Volatile
        private var INSTANCE: ApplicationController? = null
        
        fun getInstance(): ApplicationController {
            return INSTANCE ?: synchronized(this) {
                INSTANCE ?: ApplicationController().also { INSTANCE = it }
            }
        }
    }
    
    /**
     * Get all applications with optional filters
     */
    suspend fun getAllApplications(
        filters: ApplicationFilters? = null,
        page: Int = 1,
        limit: Int = 20
    ): Result<ApplicationListResponse> {
        return try {
            val supabaseClient = SupabaseClient.getInstance()
            
            // Build query with filters
            var query = "kpr_applications"
            var offset = (page - 1) * limit
            
            // Apply filters
            filters?.let { filter ->
                if (filter.status.isNotEmpty()) {
                    // In a real implementation, you'd build the query differently
                    // For now, we'll use the basic select and filter in the client
                }
            }
            
            // Get applications
            val result = supabaseClient.select(
                table = query,
                orderBy = "created_at.desc",
                limit = limit,
                offset = offset
            )
            
            val applications = result.getOrNull()?.map { app ->
                ApplicationData(
                    id = app["id"] as? String ?: "",
                    customerName = app["customer_name"] as? String ?: "",
                    customerEmail = app["customer_email"] as? String ?: "",
                    customerPhone = app["customer_phone"] as? String ?: "",
                    unitType = app["unit_type"] as? String ?: "",
                    unitBlock = app["unit_block"] as? String ?: "",
                    unitNumber = app["unit_number"] as? String ?: "",
                    status = app["status"] as? String ?: "",
                    bankId = app["bank_id"] as? String,
                    created_at = app["created_at"] as? String ?: "",
                    updated_at = app["updated_at"] as? String ?: "",
                    notes = app["notes"] as? String,
                    aging_days = calculateAging(app["created_at"] as? String ?: ""),
                    sla_status = getSLAStatus(app["created_at"] as? String ?: "", app["status"] as? String ?: "")
                )
            } ?: emptyList()
            
            // Calculate statistics
            val stats = calculateApplicationStatistics(applications)
            
            ApplicationListResponse(
                success = true,
                data = applications,
                count = applications.size,
                page = page,
                limit = limit,
                total = stats.total,
                hasMore = applications.size == limit
            )
        } catch (e: Exception) {
            Log.e(TAG, "Error fetching applications", e)
            ApplicationListResponse(
                success = false,
                data = emptyList(),
                count = 0,
                error = e.message
            )
        }
    }
    
    /**
     * Get application by ID
     */
    suspend fun getApplicationById(id: String): Result<ApplicationData> {
        return try {
            val supabaseClient = SupabaseClient.getInstance()
            val result = supabaseClient.getApplicationById(id)
            
            if (result.isSuccess) {
                val app = result.getOrNull() ?: emptyMap()
                val application = ApplicationData(
                    id = app["id"] as? String ?: "",
                    customerName = app["customer_name"] as? String ?: "",
                    customerEmail = app["customer_email"] as? String ?: "",
                    customerPhone = app["customer_phone"] as? String ?: "",
                    unitType = app["unit_type"] as? String ?: "",
                    unitBlock = app["unit_block"] as? String ?: "",
                    unitNumber = app["unit_number"] as? String ?: "",
                    status = app["status"] as? String ?: "",
                    bankId = app["bank_id"] as? String,
                    created_at = app["created_at"] as? String ?: "",
                    updated_at = app["updated_at"] as? String ?: "",
                    notes = app["notes"] as? String,
                    aging_days = calculateAging(app["created_at"] as? String ?: ""),
                    sla_status = getSLAStatus(app["created_at"] as? String ?: "", app["status"] as? String ?: "")
                )
                Result.success(application)
            } else {
                Result.failure(Exception("Application not found"))
            }
        } catch (e: Exception) {
            Log.e(TAG, "Error fetching application by ID: $id", e)
            Result.failure(e)
        }
    }
    
    /**
     * Create new application
     */
    suspend fun createApplication(request: ApplicationCreateRequest): Result<ApplicationData> {
        return try {
            val supabaseClient = SupabaseClient.getInstance()
            
            // Validate request
            val validation = validateApplicationCreation(request)
            if (!validation.isValid) {
                return Result.failure(Exception(validation.errors.join(", ")))
            }
            
            // Create application data
            val applicationData = mapOf(
                "customer_id" to request.customerId,
                "unit_id" to request.unitId,
                "bank_id" to request.bankId,
                "status" to "PENDING",
                "notes" to request.notes,
                "created_at" to getCurrentTimestamp(),
                "updated_at" to getCurrentTimestamp()
            )
            
            val result = supabaseClient.createApplication(applicationData)
            
            if (result.isSuccess) {
                val createdApp = result.getOrNull() ?: emptyMap()
                val application = ApplicationData(
                    id = createdApp["id"] as? String ?: "",
                    customerName = "", // Would fetch from customer table
                    customerEmail = "",
                    customerPhone = "",
                    unitType = "", // Would fetch from unit table
                    unitBlock = "",
                    unitNumber = "",
                    status = "PENDING",
                    bankId = request.bankId,
                    created_at = createdApp["created_at"] as? String ?: "",
                    updated_at = createdApp["updated_at"] as? String ?: "",
                    notes = request.notes,
                    aging_days = 0,
                    sla_status = "on_track"
                )
                
                // Send notification
                sendStatusChangeNotification(application, null, "PENDING")
                
                Result.success(application)
            } else {
                Result.failure(Exception("Failed to create application"))
            }
        } catch (e: Exception) {
            Log.e(TAG, "Error creating application", e)
            Result.failure(e)
        }
    }
    
    /**
     * Update application
     */
    suspend fun updateApplication(id: String, request: ApplicationUpdateRequest): Result<ApplicationData> {
        return try {
            val supabaseClient = SupabaseClient.getInstance()
            
            // Get current application
            val currentAppResult = getApplicationById(id)
            if (!currentAppResult.isSuccess) {
                return Result.failure(Exception("Application not found"))
            }
            
            val currentApp = currentAppResult.getOrNull()
            
            // Validate status change
            if (request.status != null && request.status != currentApp.status) {
                val validation = validateStatusChange(
                    currentApp.status,
                    request.status,
                    currentApp
                )
                if (!validation.isValid) {
                    return Result.failure(Exception(validation.errors.join(", ")))
                }
            }
            
            // Update application data
            val updateData = mutableMapOf<String, Any>()
            request.status?.let { updateData["status"] = it }
            request.notes?.let { updateData["notes"] = it }
            request.bankId?.let { updateData["bank_id"] = it }
            updateData["updated_at"] = getCurrentTimestamp()
            
            val result = supabaseClient.updateApplication(id, updateData)
            
            if (result.isSuccess) {
                val updatedApp = result.getOrNull() ?: emptyMap()
                val application = ApplicationData(
                    id = id,
                    customerName = currentApp.customerName,
                    customerEmail = currentApp.customerEmail,
                    customerPhone = currentApp.customerPhone,
                    unitType = currentApp.unitType,
                    unitBlock = currentApp.unitBlock,
                    unitNumber = currentApp.unitNumber,
                    status = request.status ?: currentApp.status,
                    bankId = request.bankId ?: currentApp.bankId,
                    created_at = currentApp.created_at,
                    updated_at = updatedApp["updated_at"] as? String ?: currentApp.updated_at,
                    notes = request.notes ?: currentApp.notes,
                    aging_days = calculateAging(currentApp.created_at),
                    sla_status = getSLAStatus(currentApp.created_at, request.status ?: currentApp.status)
                )
                
                // Log status change
                if (request.status != null && request.status != currentApp.status) {
                    logStatusChange(id, currentApp.status, request.status, request.notes)
                    
                    // Send notification
                    sendStatusChangeNotification(application, currentApp.status, request.status)
                }
                
                Result.success(application)
            } else {
                Result.failure(Exception("Failed to update application"))
            }
        } catch (e: Exception) {
            Log.e(TAG, "Error updating application", e)
            Result.failure(e)
        }
    }
    
    /**
     * Delete application
     */
    suspend fun deleteApplication(id: String): Result<Unit> {
        return try {
            val supabaseClient = SupabaseClient.getInstance()
            
            // Get current application
            val currentAppResult = getApplicationById(id)
            if (!currentAppResult.isSuccess) {
                return Result.failure(Exception("Application not found"))
            }
            
            val currentApp = currentAppResult.getOrNull()
            
            // Check if can be deleted
            if (!canDeleteApplication(currentApp.status)) {
                return Result.failure(Exception("Application cannot be deleted on this status"))
            }
            
            // Delete application
            val result = supabaseClient.deleteApplication(id)
            
            if (result.isSuccess) {
                // Update unit status if needed
                // This would require unit management logic
                Result.success(Unit)
            } else {
                Result.failure(Exception("Failed to delete application"))
            }
        } catch (e: Exception) {
            Log.e(TAG, "Error deleting application", e)
            Result.failure(e)
        }
    }
    
    /**
     * Get applications by status
     */
    suspend fun getApplicationsByStatus(
        status: String,
        page: Int = 1,
        limit: Int = 20
    ): Result<ApplicationListResponse> {
        val filters = ApplicationFilters(status = listOf(status))
        return getAllApplications(filters, page, limit)
    }
    
    /**
     * Get application statistics
     */
    suspend fun getApplicationStatistics(): Result<ApplicationStatistics> {
        return try {
            val supabaseClient = SupabaseClient.getInstance()
            val result = supabaseClient.getApplications()
            
            if (result.isSuccess) {
                val applications = result.getOrNull() ?: emptyList()
                val stats = calculateApplicationStatistics(applications)
                Result.success(stats)
            } else {
                Result.failure(Exception("Failed to fetch application statistics"))
            }
        } catch (e: Exception) {
            Log.e(TAG, "Error fetching application statistics", e)
            Result.failure(e)
        }
    }
    
    /**
     * Real-time application updates
     */
    fun observeApplicationUpdates(): Flow<ApplicationData> {
        return flow {
            val supabaseClient = SupabaseClient.getInstance()
            
            supabaseClient.subscribeToApplications()
                .collect { message ->
                    // Parse real-time message and emit application data
                    try {
                        val applicationId = message.payload["id"] as? String
                        if (applicationId != null) {
                            // Fetch updated application
                            val appResult = getApplicationById(applicationId)
                            if (appResult.isSuccess) {
                                emit(appResult.getOrNull())
                            }
                        }
                    } catch (e: Exception) {
                        Log.e(TAG, "Error processing real-time update", e)
                    }
                }
        }
    }
    
    // Private helper methods
    private fun validateApplicationCreation(request: ApplicationCreateRequest): ValidationResult {
        val errors = mutableListOf<String>()
        
        // Validate customer exists (in real implementation)
        if (request.customerId.isEmpty()) {
            errors.add("Customer ID is required")
        }
        
        // Validate unit exists (in real implementation)
        if (request.unitId.isEmpty()) {
            errors.add("Unit ID is required")
        }
        
        // Validate bank if provided
        if (request.bankId?.isEmpty() == false) {
            // In real implementation, validate bank exists
        }
        
        return ValidationResult(
            isValid = errors.isEmpty(),
            errors = errors
        )
    }
    
    private fun validateStatusChange(
        fromStatus: String,
        toStatus: String,
        application: ApplicationData
    ): ValidationResult {
        val errors = mutableListOf<String>()
        
        // Define valid status transitions
        val validTransitions = mapOf(
            "PENDING" to listOf("DOCUMENT_COLLECTION", "CANCELLED"),
            "DOCUMENT_COLLECTION" to listOf("BANK_SUBMISSION", "CANCELLED"),
            "BANK_SUBMISSION" to listOf("BANK_VERIFICATION", "CANCELLED"),
            "BANK_VERIFICATION" to listOf("BANK_APPROVAL", "REJECTED", "CANCELLED"),
            "BANK_APPROVAL" to listOf("APPROVED", "CANCELLED"),
            "APPROVED" to listOf("COMPLETED", "CANCELLED"),
            "COMPLETED" to listOf("BAST_COMPLETED"),
            "BAST_COMPLETED" to listOf(),
            "REJECTED" to listOf(),
            "CANCELLED" to listOf()
        )
        
        val allowedStatuses = validTransitions[fromStatus] ?: emptyList()
        if (!allowedStatuses.contains(toStatus)) {
            errors.add("Status change from $fromStatus to $toStatus is not allowed")
        }
        
        return ValidationResult(
            isValid = errors.isEmpty(),
            errors = errors
        )
    }
    
    private fun canDeleteApplication(status: String): Boolean {
        return listOf("PENDING", "DOCUMENT_COLLECTION").contains(status)
    }
    
    private fun calculateAging(createdDate: String): Int {
        return try {
            val created = java.time.Instant.parse(createdDate)
            val now = java.time.Instant.now()
            val days = java.time.Duration.between(created, now).days
            days.toInt()
        } catch (e: Exception) {
            0
        }
    }
    
    private fun getSLADaysByStatus(status: String): Int {
        return when (status) {
            "PENDING" -> 3
            "DOCUMENT_COLLECTION" -> 7
            "BANK_SUBMISSION" -> 2
            "BANK_VERIFICATION" -> 14
            "BANK_APPROVAL" -> 3
            "APPROVED" -> 30
            "COMPLETED" -> 0
            "BAST_COMPLETED" -> 0
            "REJECTED" -> 0
            "CANCELLED" -> 0
            else -> 30
        }
    }
    
    private fun getSLAStatus(createdDate: String, status: String): String {
        val slaDays = getSLADaysByStatus(status)
        val daysUsed = calculateAging(createdDate)
        
        return when {
            daysUsed <= slaDays -> "on_track"
            daysUsed <= slaDays + 7 -> "warning"
            daysUsed > slaDays + 7 -> "overdue"
            else -> "critical"
        }
    }
    
    private fun calculateApplicationStatistics(applications: List<ApplicationData>): ApplicationStatistics {
        val statusCounts = applications.groupBy { it.status }
        
        return ApplicationStatistics(
            total = applications.size,
            byStatus = statusCounts.mapValues { it.size },
            byBank = emptyMap(), // Would need to implement
            byMonth = emptyMap(), // Would need to implement
            averageProcessingTime = 0, // Would need to implement
            approvalRate = 0.0, // Would need to implement
            rejectionRate = 0.0  // Would need to implement
        )
    }
    
    private fun logStatusChange(
        applicationId: String,
        fromStatus: String,
        toStatus: String,
        notes: String?
    ) {
        Log.d(TAG, "Status changed for application $applicationId: $fromStatus -> $toStatus")
        // In real implementation, this would log to audit trail
    }
    
    private fun sendStatusChangeNotification(
        application: ApplicationData,
        fromStatus: String?,
        toStatus: String
    ) {
        Log.d(TAG, "Sending status change notification for application ${application.id}")
        // In real implementation, this would send notification
    }
    
    private fun getCurrentTimestamp(): String {
        return java.time.Instant.now().toString()
    }
}

// Data classes
data class ApplicationFilters(
    val status: List<String> = emptyList(),
    val customerId: String? = null,
    val unitId: String? = null,
    val bankId: String? = null,
    val dateFrom: String? = null,
    val dateTo: String? = null,
    val search: String? = null
)

data class ApplicationCreateRequest(
    val customerId: String,
    val unitId: String,
    val bankId: String? = null,
    val notes: String? = null
)

data class ApplicationUpdateRequest(
    val status: String? = null,
    val notes: String? = null,
    val bankId: String? = null
)

data class ApplicationData(
    val id: String,
    val customerName: String,
    val customerEmail: String,
    val customerPhone: String,
    val unitType: String,
    val unitBlock: String,
    val unitNumber: String,
    val status: String,
    val bankId: String?,
    val created_at: String,
    val updated_at: String,
    val notes: String?,
    val aging_days: Int,
    val sla_status: String
)

data class ApplicationListResponse(
    val success: Boolean,
    val data: List<ApplicationData>,
    val count: Int,
    val page: Int,
    val limit: Int,
    val total: Int = 0,
    val hasMore: Boolean = false,
    val error: String? = null
)

data class ApplicationStatistics(
    val total: Int,
    val byStatus: Map<String, Int>,
    val byBank: Map<String, Int>,
    val byMonth: Map<String, Int>,
    val averageProcessingTime: Int,
    val approvalRate: Double,
    val rejectionRate: Double
)

data class ValidationResult(
    val isValid: Boolean,
    val errors: List<String>
)
